build/makefiles/provider.mk

```diff
@@ -0,0 +1,198 @@
+endif
+
+par_targets ?= \
+	x86_64-unknown-linux-gnu \
+   	x86_64-apple-darwin \
+	armv7-unknown-linux-gnueabihf \
+   	aarch64-unknown-linux-gnu \
+   	aarch64-apple-darwin \
+   	x86_64-pc-windows-gnu
+
+bin_targets = $(foreach target,$(par_targets),target/$(target)/release/$(bin_name))
+
+# pick target0 for starting par
+ifeq ($(machine_id)-$(platform_id),x86_64-Linux)
+	par_target0 = x86_64-unknown-linux-gnu
+else
+	ifeq ($(machine_id)-$(platform_id),x86_64-Darwin)
+		par_target0 = x86_64-apple-darwin
+	else
+		# default to x86_64-linux
+		par_target0=x86_64-unknown-linux-gnu
+	endif
+endif
+
+# the target of the current platform, as defined by cross
+cross_target0=target/$(par_target0)/release/$(bin_name)
+# bin_target0=$(cross_target0)
+bin_target0=target/release/$(bin_name)
+
+# traverse subdirs
+.ONESHELL:
+ifneq ($(subdirs),)
+$(top_targets)::
+	for dir in $(subdirs); do \
+		$(MAKE) -C $$dir $@ weld=$(weld); \
+	done
+endif
+
+# this target must be listed first so that including this makefile
+# doesn't trigger other rules
+#all::
+
+# build par file for current platform
+par: release $(dest_par) build/stub.exs
+
+# rebuild base par if target0 changes
+$(dest_par): $(bin_target0) Makefile
+	@mkdir -p $(dir $(dest_par))
+	rm -f $@
+	par_arch=`echo $(par_target0) | sed -E 's/([^-]+)-([^-]+)-([^-]+)(-gnu)?/\1-\3/'`
+	wash par create \
+		--arch $$par_arch \
+		--binary $(bin_target0) \
+		--capid $(CAPABILITY_ID) \
+		--name $(NAME) \
+		--vendor $(VENDOR) \
+		--version $(VERSION) \
+		--revision $(REVISION) \
+		--destination $@ \
+		--compress
+	@echo Created $@
+
+
+# par-full adds all the other targets to the base par
+par-full: $(dest_par) $(bin_targets)
+	# add other defined targets
+	for target in $(par_targets); do \
+	    target_dest=target/$$target/release/$(bin_name);  \
+	    par_arch=`echo -n $$target | sed -E 's/([^-]+)-([^-]+)-([^-]+)(-gnu)?/\1-\3/'`; \
+		echo building $$par_arch; \
+		if [ $${target_dest} != $(cross_target0) ] && [ -f $$target_dest ]; then \
+		    wash par insert --arch $$par_arch --binary $$target_dest $@; \
+		fi; \
+	done
+
+# create rust build targets
+ifeq ($(wildcard ./Cargo.toml),./Cargo.toml)
+
+# rust dependencies
+RUST_DEPS += $(wildcard src/*.rs) $(wildcard target/release/deps/*) Cargo.toml Makefile
+
+target/release/$(bin_name): $(RUST_DEPS)
+	cargo build --release
+	@rm -f $(dest_par)
+
+target/debug/$(bin_name): $(RUST_DEPS)
+	cargo build
+	@rm -f $(dest_par)
+
+target/%/release/$(bin_name): $(RUST_DEPS)
+	tname=`echo -n $@ | sed -E 's_target/([^/]+)/release.*$$_\1_'` &&\
+	cross build --release --target $$tname
+	@rm -f $(dest_par)
+
+endif
+
+# make exs stub for provider
+build/stub.exs: $(dest_par)
+	@mkdir -p $(dir $@)
+	@cat <<- EOF > $@
+	%{
+		name: $(NAME),
+		path: "$(abspath $(dest_par))",
+		key: "$(shell wash par inspect $(dest_par) -o json | jq -r ".service")",
+		link: "default",
+		contract: "$(CAPABILITY_ID)",
+	},
+	EOF
+
+
+# is this needed for cap providers?
```

Comment:
If this is commented out because it's not needed, then we should remove these lines from the file.

Reply:
deleted. weld commands moving to wash.